File: TypeScript/PaymentsService/src/Domain/Payments/PaymentProcessor.ts
import { IBaseLogger } from "../Logging/IBaseLogger";
import { IStatisticsService } from "../Statistics/IStatisticsService";
import { IPaymentAbstractFactory } from "./IPaymentAbstractFactory";
import { ProcessPaymentBaseRequest } from "./ProcessPaymentBaseRequest";
import { ProcessPaymentBaseResponse } from "./ProcessPaymentBaseResponse";


export class PaymentProcessor {
    constructor(private factory: IPaymentAbstractFactory,
        private statisticsService: IStatisticsService,
        private logger: IBaseLogger) { }

    processPayment(request: ProcessPaymentBaseRequest): ProcessPaymentBaseResponse {
        this.logger.log("Processing request", request);

        const validator = this.factory.getProcessPaymentRequestValidator();

        const validationResult = validator.validateRequest(request);

        if(!validationResult.isValid)
        {
            this.logger.log("Validation failed");
            return this.factory.getValidationFailedResponse(validationResult);
        }

        const client = this.factory.getPaymentClient();

        const paymentResult: ProcessPaymentBaseResponse = client.processPayment(request);

        this.statisticsService.countStatistics(request, paymentResult);

        this.logger.log("Processing request finished with result", paymentResult);

        return paymentResult;
    }
}

File: Python/PaymentsService/Domain/Payments/PaymentProcessor.py
from Domain.Logging.IBaseLogger import IBaseLogger
from Domain.Payments.IPaymentAbstractFactory import IPaymentAbstractFactory
from Domain.Payments.ProcessPaymentBaseRequest import ProcessPaymentBaseRequest
from Domain.Payments.ProcessPaymentBaseResponse import ProcessPaymentBaseResponse
from Domain.Statistics.IStatisticsService import IStatisticsService


class PaymentProcessor:
    def __init__(self, factory: IPaymentAbstractFactory, statistics_service: IStatisticsService, logger: IBaseLogger):
        self._factory = factory
        self._statistics_service = statistics_service
        self._logger = logger

    def process_payment(self, request: ProcessPaymentBaseRequest) -> ProcessPaymentBaseResponse:
        self._logger.log_message_and_object("Processing request", request)

        validator = self._factory.get_process_payment_request_validator()

        validation_result = validator.validate_request(request)

        if not validation_result.is_valid:
            self._logger.log("Validation failed")
            return self._factory.get_validation_failed_response(validation_result)

        client = self._factory.get_payment_client()

        payment_result = client.process_payment(request)

        self._statistics_service.count_statistics(request, payment_result)

        self._logger.log_message_and_object("Processing request finished with result", payment_result)

        return payment_result
